### PR TITLE
Introduce config to cache segments locally before uncompressing

### PR DESCRIPTION
**Pull request by niketh**

### CODE REVIEW

**Comment by fjy** [comment]
Just a general comment overall to discuss, but I'm starting to dislike having so many 'optional' features and increasingly adding more configuration. It is just becoming difficult to maintain and understand what is important to turn on when. I think Druid should strive to have less configuration, not more.

**Comment by niketh** [comment]
@fjy I agree. But in this specific case, S3DataSegmentPuller always downloads to a temp directory before it can uncompress. This behaviour might not be required for everyone

File: docs/content/configuration/historical.md [hunk 1]
```diff
@@ -37,6 +37,7 @@ The historical node uses several of the global configs in [Configuration](../con
 |`druid.segmentCache.announceIntervalMillis`|How frequently to announce segments while segments are loading from cache. Set this value to zero to wait for all segments to be loaded before announcing.|5000 (5 seconds)|
 |`druid.segmentCache.numLoadingThreads`|How many segments to load concurrently from from deep storage.|1|
 |`druid.segmentCache.numBootstrapThreads`|How many segments to load concurrently from local storage at startup.|1|
+|`druid.segmentCache.cacheSegmentsLocally`|Download segments to temp before uncompressing.|false|
```
Comment (gianm) [inline docs/content/configuration/historical.md]: I'm not sure what this means or when I would want to set it, which I think means it should be written more clearly.

**Comment by niketh** [comment]
@gianm Is the documentation clearer now?

**Comment by drcrallen** [comment]
I don't think the API for load spec should change.

This assumes all pullers will always require a copy and a processing of the data, which is not true. It is possible to have a load spec that wires up data without needing to copy/process the data in the way that the s3 puller does currently.

As such caching is a puller property, not a LoadSpec property.

**Comment by cheddar** [comment]
I agree with @drcrallen if we add this config, it should be through adding a Config object to the S3DataSegmentPuller's constructor without any change to the interfaces.

Alternatively, taking what @fjy said (which was parrotted by @himanshug internally), we could just eliminate the download-first strategy altogether.  @drcrallen you guys use the S3 puller, would you be against eliminating download-first altogether?

**Comment by niketh** [comment]
Closed in favour of https://github.com/druid-io/druid/pull/4364

**Pull request closed without merging**